node/forest_libp2p/src/blocksync/provider.rs

```diff
@@ -0,0 +1,241 @@
+};
+
+/// Builds blocksync response out of chain data.
+pub fn make_blocksync_response<DB>(db: Arc<DB>, request: &BlockSyncRequest) -> BlockSyncResponse
+where
+    DB: BlockStore,
+{
+    let mut response_chain: Vec<TipsetBundle> = vec![];
+
+    let mut curr_tipset_cids = request.start.clone();
+
+    loop {
+        let mut tipset_bundle: TipsetBundle = Default::default();
+        let tipset = match chain::tipset_from_keys(db.as_ref(), &TipsetKeys::new(curr_tipset_cids))
+        {
+            Ok(tipset) => tipset,
+            Err(err) => {
+                debug!("Cannot get tipset from keys: {}", err);
+
+                return BlockSyncResponse {
+                    chain: vec![],
+                    status: BlockSyncResponseStatus::InternalError,
+                    message: "Tipset was not found in the database".to_owned(),
+                };
+            }
+        };
+
+        if request.include_blocks() {
+            tipset_bundle.blocks = tipset.blocks().to_vec();
+        }
+
+        if request.include_messages() {
+            match compact_messages(db.as_ref(), &tipset) {
+                Ok(compacted_messages) => tipset_bundle.messages = Some(compacted_messages),
+                Err(err) => {
+                    debug!("Cannot compact messages for tipset: {}", err);
+
+                    return BlockSyncResponse {
+                        chain: vec![],
+                        status: BlockSyncResponseStatus::InternalError,
+                        message: "Can not fullfil the request".to_owned(),
+                    };
+                }
+            }
+        }
+
+        response_chain.push(tipset_bundle);
+
+        if response_chain.len() as u64 >= request.request_len || tipset.epoch() == 0 {
+            break;
+        }
+
+        curr_tipset_cids = tipset.parents().cids().to_vec();
+    }
+
+    let result_chain_length = response_chain.len() as u64;
+
+    BlockSyncResponse {
+        chain: response_chain,
+        status: if result_chain_length < request.request_len {
+            BlockSyncResponseStatus::PartialResponse
+        } else {
+            BlockSyncResponseStatus::Success
+        },
+        message: "Success".to_owned(),
+    }
+}
+
+// Builds CompactedMessages for given Tipset.
+fn compact_messages<DB>(db: &DB, tipset: &Tipset) -> Result<CompactedMessages, ChainError>
+where
+    DB: BlockStore,
+{
+    let mut bls_messages_order = HashMap::new();
+    let mut secp_messages_order = HashMap::new();
+    let mut bls_cids_combined: Vec<Cid> = vec![];
+    let mut secp_cids_combined: Vec<Cid> = vec![];
+    let mut bls_msg_includes: Vec<Vec<u64>> = vec![];
+    let mut secp_msg_includes: Vec<Vec<u64>> = vec![];
+
+    for block_header in tipset.blocks().iter() {
+        let (bls_cids, secp_cids) = chain::read_msg_cids(db, block_header.messages())?;
+
+        let mut block_include = vec![];
+        let mut secp_include = vec![];
```

Comment:
can you set the capacity to the length of the cids read above?

Reply:
sure, updated.